File: aricode/sparring/challenges/11_isqrt/isqrt.c
/* C - Challenge 11: Integer Square Root (Newton's Method) */
/* Compute isqrt(16129) = 127 */

int isqrt(int n) {
    if (n < 2) return n;
    int x = n;
    int y = (x + 1) / 2;
    while (y < x) {
        x = y;
        y = (x + n / x) / 2;
    }
    return x;
}

int main(void) {
    return isqrt(16129);
}

File: aricode/sparring/challenges/11_isqrt/isqrt.rs
// Rust - Challenge 11: Integer Square Root (Newton's Method)
// Compute isqrt(16129) = 127
use std::process::ExitCode;

fn isqrt(n: i32) -> i32 {
    if n < 2 { return n; }
    let mut x = n;
    let mut y = (x + 1) / 2;
    while y < x {
        x = y;
        y = (x + n / x) / 2;
    }
    x
}

fn main() -> ExitCode {
    ExitCode::from(isqrt(16129) as u8)
}
